File: backend/src/services/auth.service.ts
import { ADMIN_REGISTRATION_SECRET } from "../config/constants.js";
import { RefreshToken } from "../models/refresh-token.model.js";
import { User, type IUser } from "../models/user.model.js";
import type {
  AdminRegisterInput,
  AuthUserResponse,
  LoginInput,
  LoginResponseData,
  LogoutResponseData,
  RefreshTokenInput,
  RefreshTokenResponseData,
  RegisterInput,
  RegisterResponseData,
} from "../types/auth.types.js";
import {
  createAuthTokenPair,
  hashRefreshToken,
  verifyRefreshToken,
} from "../utils/jwt.js";
import { clearCartForUser, mergeGuestCartIntoUserCart } from "./cart.service.js";

export class AuthServiceError extends Error {
  statusCode: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = "AuthServiceError";
    this.statusCode = statusCode;
  }
}

function normalizeEmail(email: string): string {
  return email.trim().toLowerCase();
}

function isDuplicateKeyError(error: unknown): boolean {
  return (
    typeof error === "object" &&
    error !== null &&
    "code" in error &&
    error.code === 11000
  );
}

export function toAuthUserResponse(user: IUser): AuthUserResponse {
  return {
    id: user._id.toString(),
    email: user.email,
    name: user.name,
    ...(user.phone === undefined ? {} : { phone: user.phone }),
    ...(user.avatar === undefined ? {} : { avatar: user.avatar }),
    role: user.role,
    status: user.status,
    isActive: user.isActive !== false,
  };
}

async function registerUserByRole(
  input: RegisterInput,
  role: IUser["role"],
): Promise<RegisterResponseData> {
  const email = normalizeEmail(input.email);
  const existingUser = await User.exists({ email });

  if (existingUser !== null) {
    throw new AuthServiceError("Email is already registered", 409);
  }

  try {
    const user = await User.create({
      email,
      password: input.password,
      name: input.name.trim(),
      role,
      ...(input.phone === undefined ? {} : { phone: input.phone }),
    });

    return { user: toAuthUserResponse(user) };
  } catch (error) {
    if (isDuplicateKeyError(error) === true) {
      throw new AuthServiceError("Email is already registered", 409);
    }

    throw error;
  }
}

export async function registerUser(input: RegisterInput): Promise<RegisterResponseData> {
  return registerUserByRole(input, "customer");
}

export async function registerAdmin(
  input: AdminRegisterInput,
): Promise<RegisterResponseData> {
  if (
    ADMIN_REGISTRATION_SECRET === undefined ||
    ADMIN_REGISTRATION_SECRET.trim() === ""
  ) {
    throw new AuthServiceError("Admin registration is not configured", 403);
  }

  if (input.adminSecret !== ADMIN_REGISTRATION_SECRET) {
    throw new AuthServiceError("Admin registration secret is invalid", 403);
  }

  return registerUserByRole(input, "admin");
}

export async function loginUser(input: LoginInput): Promise<LoginResponseData> {
  const email = normalizeEmail(input.email);
  const user = await User.findOne({ email }).select("+password");

  if (user === null) {
    throw new AuthServiceError("Email or password is incorrect", 401);
  }

  const passwordMatches = await user.matchPassword(input.password);

  if (passwordMatches === false) {
    throw new AuthServiceError("Email or password is incorrect", 401);
  }

  if (user.status === "banned" || user.isActive === false) {
    throw new AuthServiceError("This account has been banned", 403);
  }

  const tokens = createAuthTokenPair(user);

  await RefreshToken.create({
    userId: user._id,
    tokenHash: tokens.refreshTokenHash,
    expiresAt: tokens.refreshTokenExpiresAt,
  });

  return {
    user: toAuthUserResponse(user),
    accessToken: tokens.accessToken,
    refreshToken: tokens.refreshToken,
  };
}

export async function loginAdmin(input: LoginInput): Promise<LoginResponseData> {
  const response = await loginUser(input);

  if (response.user.role !== "admin") {
    throw new AuthServiceError("Admin account is required", 403);
  }

  return response;
}

function invalidRefreshTokenError(): AuthServiceError {
  return new AuthServiceError("Refresh token is invalid or expired", 401);
}

export async function refreshAccessToken(
  input: RefreshTokenInput,
): Promise<RefreshTokenResponseData> {
  let payload;

  try {
    payload = verifyRefreshToken(input.refreshToken);
  } catch {
    throw invalidRefreshTokenError();
  }

  const storedRefreshToken = await RefreshToken.findOneAndUpdate(
    {
      userId: payload.sub,
      tokenHash: hashRefreshToken(input.refreshToken),
      expiresAt: { $gt: new Date() },
      revokedAt: { $exists: false },
    },
    {
      $set: {
        revokedAt: new Date(),
      },
    },
    {
      new: true,
    },
  ).select("+tokenHash");

  if (storedRefreshToken === null) {
    await RefreshToken.deleteMany({
      userId: payload.sub,
      revokedAt: { $exists: false },
    });
    throw invalidRefreshTokenError();
  }

  const user = await User.findById(payload.sub);

  if (user === null || user.status === "banned" || user.isActive === false) {
    await RefreshToken.deleteMany({ userId: storedRefreshToken.userId });
    throw invalidRefreshTokenError();
  }

  const tokens = createAuthTokenPair(user);

  await RefreshToken.create({
    userId: user._id,
    tokenHash: tokens.refreshTokenHash,
    expiresAt: tokens.refreshTokenExpiresAt,
  });

  return {
    accessToken: tokens.accessToken,
    refreshToken: tokens.refreshToken,
  };
}

export async function logoutSession(
  input: RefreshTokenInput,
): Promise<LogoutResponseData> {
  try {
    const payload = verifyRefreshToken(input.refreshToken);
    const refreshTokenDeletion = await RefreshToken.deleteOne({
      tokenHash: hashRefreshToken(input.refreshToken),
    });

    if (refreshTokenDeletion.deletedCount === 1) {
      await clearCartForUser(payload.sub);
    }
  } catch {
    await RefreshToken.deleteOne({
      tokenHash: hashRefreshToken(input.refreshToken),
    });
  }

  return {
    message: "Logged out",
  };
}

export async function mergeGuestCartForUser(
  userId: string,
  guestId: string | undefined,
): Promise<boolean> {
  if (guestId === undefined || guestId.trim() === "") {
    return false;
  }

  return mergeGuestCartIntoUserCart(userId, guestId);
}

export async function getCurrentAuthUser(
  userId: string,
): Promise<AuthUserResponse> {
  const user = await User.findById(userId);

  if (user === null || user.status !== "active" || user.isActive === false) {
    throw new AuthServiceError("User not found", 404);
  }

  return toAuthUserResponse(user);
}

File: backend/src/main/java/com/season/app/service/AuthService.java
package com.season.app.service;

import com.season.app.dto.auth.*;
import com.season.app.model.RefreshToken;
import com.season.app.model.User;
import com.season.app.repository.RefreshTokenRepository;
import com.season.app.repository.UserRepository;
import com.season.app.util.JwtUtil;
import io.jsonwebtoken.Claims;
import java.time.Instant;
import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class AuthService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private RefreshTokenRepository refreshTokenRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtUtil jwtUtil;

    @Autowired
    private CartService cartService;

    private final String adminRegistrationSecret;

    public AuthService(
            UserRepository userRepository,
            @Value("${app.admin.registration-secret:}") String adminRegistrationSecret
    ) {
        this.userRepository = userRepository;
        this.adminRegistrationSecret = adminRegistrationSecret;
    }

    public RegisterResponseData registerUser(RegisterInput input) {
        return registerUserByRole(input, "customer");
    }

    public RegisterResponseData registerAdmin(RegisterInput input, String providedAdminSecret) {
        if (adminRegistrationSecret == null || adminRegistrationSecret.trim().isEmpty()) {
            throw new AuthServiceError("Admin registration is not configured", HttpStatus.FORBIDDEN.value());
        }
        if (providedAdminSecret == null || !providedAdminSecret.equals(adminRegistrationSecret)) {
            throw new AuthServiceError("Admin registration secret is invalid", HttpStatus.FORBIDDEN.value());
        }
        return registerUserByRole(input, "admin");
    }

    public LoginResponseData loginUser(LoginInput input) {
        String email = normalizeEmail(input.getEmail());
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new AuthServiceError("Email or password is incorrect", HttpStatus.UNAUTHORIZED.value()));

        if (!passwordEncoder.matches(input.getPassword(), user.getPassword())) {
            throw new AuthServiceError("Email or password is incorrect", HttpStatus.UNAUTHORIZED.value());
        }

        if ("banned".equals(user.getStatus()) || Boolean.FALSE.equals(user.getIsActive())) {
            throw new AuthServiceError("This account has been banned", HttpStatus.FORBIDDEN.value());
        }

        String accessToken = jwtUtil.generateAccessToken(user);
        String refreshToken = jwtUtil.generateRefreshToken(user);
        String refreshTokenHash = JwtUtil.hashRefreshToken(refreshToken);

        RefreshToken rt = RefreshToken.builder()
                .userId(user.getId())
                .tokenHash(refreshTokenHash)
                .expiresAt(jwtUtil.validateRefreshToken(refreshToken).getExpiration().toInstant())
                .build();
        refreshTokenRepository.save(rt);

        LoginResponseData response = new LoginResponseData();
        response.setUser(toAuthUserResponse(user));
        response.setAccessToken(accessToken);
        response.setRefreshToken(refreshToken);

        return response;
    }

    public RefreshTokenResponseData refreshAccessToken(RefreshTokenInput input) {
        Claims claims;
        try {
            claims = jwtUtil.validateRefreshToken(input.getRefreshToken());
        } catch (Exception e) {
            throw new AuthServiceError("Refresh token is invalid or expired", HttpStatus.UNAUTHORIZED.value());
        }

        String userId = claims.getSubject();
        String tokenHash = JwtUtil.hashRefreshToken(input.getRefreshToken());

        // Tìm token cũ, thu hồi (revoke) để xoay vòng token bảo mật
        Optional<RefreshToken> storedTokenOpt = refreshTokenRepository.findByTokenHash(tokenHash);
        if (storedTokenOpt.isEmpty() || storedTokenOpt.get().getRevokedAt() != null || storedTokenOpt.get().getExpiresAt().isBefore(Instant.now())) {
            // Nghi vấn bị tấn công giả mạo token -> Xoá tất cả token cũ của user
            refreshTokenRepository.deleteByUserId(userId);
            throw new AuthServiceError("Refresh token is invalid or expired", HttpStatus.UNAUTHORIZED.value());
        }

        RefreshToken storedToken = storedTokenOpt.get();
        storedToken.setRevokedAt(Instant.now());
        refreshTokenRepository.save(storedToken);

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new AuthServiceError("User not found", HttpStatus.NOT_FOUND.value()));

        if ("banned".equals(user.getStatus()) || Boolean.FALSE.equals(user.getIsActive())) {
            refreshTokenRepository.deleteByUserId(userId);
            throw new AuthServiceError("Refresh token is invalid or expired", HttpStatus.UNAUTHORIZED.value());
        }

        String newAccessToken = jwtUtil.generateAccessToken(user);
        String newRefreshToken = jwtUtil.generateRefreshToken(user);
        String newRefreshTokenHash = JwtUtil.hashRefreshToken(newRefreshToken);

        RefreshToken nextToken = RefreshToken.builder()
                .userId(user.getId())
                .tokenHash(newRefreshTokenHash)
                .expiresAt(jwtUtil.validateRefreshToken(newRefreshToken).getExpiration().toInstant())
                .build();
        refreshTokenRepository.save(nextToken);

        RefreshTokenResponseData response = new RefreshTokenResponseData();
        response.setAccessToken(newAccessToken);
        response.setRefreshToken(newRefreshToken);
        return response;
    }

    public LogoutResponseData logoutSession(RefreshTokenInput input) {
        try {
            Claims claims = jwtUtil.validateRefreshToken(input.getRefreshToken());
            String userId = claims.getSubject();
            String tokenHash = JwtUtil.hashRefreshToken(input.getRefreshToken());

            Optional<RefreshToken> rtOpt = refreshTokenRepository.findByTokenHash(tokenHash);
            if (rtOpt.isPresent()) {
                refreshTokenRepository.delete(rtOpt.get());
                cartService.clearCartForUser(userId);
            }
        } catch (Exception e) {
            String tokenHash = JwtUtil.hashRefreshToken(input.getRefreshToken());
            refreshTokenRepository.findByTokenHash(tokenHash).ifPresent(rt -> refreshTokenRepository.delete(rt));
        }

        LogoutResponseData response = new LogoutResponseData();
        response.setMessage("Logged out");
        return response;
    }

    public AuthUserResponse getCurrentAuthUser(String userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new AuthServiceError("User not found", HttpStatus.NOT_FOUND.value()));

        if (Boolean.FALSE.equals(user.getIsActive()) || !"active".equals(user.getStatus())) {
            throw new AuthServiceError("User not found", HttpStatus.NOT_FOUND.value());
        }

        return toAuthUserResponse(user);
    }

    private RegisterResponseData registerUserByRole(RegisterInput input, String role) {
        String email = normalizeEmail(input.getEmail());
        Optional<User> existingUser = userRepository.findByEmail(email);
        if (existingUser.isPresent()) {
            throw new AuthServiceError("Email is already registered", HttpStatus.CONFLICT.value());
        }

        User user = User.builder()
                .email(email)
                .password(passwordEncoder.encode(input.getPassword()))
                .name(input.getName() == null ? null : input.getName().trim())
                .phone(input.getPhone())
                .role(role)
                .build();

        User savedUser = userRepository.save(user);

        RegisterResponseData response = new RegisterResponseData();
        response.setUser(toAuthUserResponse(savedUser));
        return response;
    }

    private String normalizeEmail(String email) {
        return email == null ? "" : email.trim().toLowerCase();
    }

    private AuthUserResponse toAuthUserResponse(User user) {
        AuthUserResponse response = new AuthUserResponse();
        response.setId(user.getId());
        response.setEmail(user.getEmail());
        response.setName(user.getName());
        response.setPhone(user.getPhone());
        response.setAvatar(user.getAvatar());
        response.setRole(user.getRole());
        response.setStatus(user.getStatus());
        response.setIsActive(Boolean.FALSE.equals(user.getIsActive()) == false);
        return response;
    }

    public static class AuthServiceError extends RuntimeException {
        private final int statusCode;
        public AuthServiceError(String message, int statusCode) {
            super(message);
            this.statusCode = statusCode;
        }
        public int getStatusCode() { return statusCode; }
    }
}